idl/read_command.idl.hh

```diff
@@ -50,6 +50,8 @@ class partition_slice {
 struct max_result_size {
     uint64_t soft_limit;
     uint64_t hard_limit;
+    uint64_t unlimited_soft_limit [[version 4.7]] = 0;
+    uint64_t unlimited_hard_limit [[version 4.7]] = 0;
```

Comment:
We don't need 4 limits, we need only 3:
* normal_limit (page size)
* unlimited_soft_limit
* unlimited_hard_limit

I think we can get away by adding just one new limit. Since the existing ones are already called `{soft/hard}_limit`, we should add a third called `page_size`.
The current interface of these limits is as follows:
* For all normal queries `soft_limit` == `hard_limit` and the value is used as the size of the page.
* For unlimited queries, we warn after reaching the soft limit and fail after reaching the hard one.

While in a mixed cluster, the new interface would work as follows:
1. Old coordinator/new replica, normal query -> just like now
2. Old coordinator/new replica, unpaged query -> just like now
2. Old coordinator/new replica, reverse query -> use limits as soft/hard limit like now, fill page_size from local config (just a constant actually)

With a new coordinator/new replica, all would work normally, all limits interpreted according to their names.

Reply:
BTW. do these `[[version ...]]` marks actually do something or do they just serve as a comment?

Reply:
A comment as far as I understand.

Reply:
I don't follow @denesb. New replicas need to have special soft and hard limits for reversing reader and those limits should be much higher than regular soft and hard limits. Are you proposing to make soft_limit a page size for readers that implement reversing natively and to make hard_limit a page size for reversing reader?

Reply:
This `max_result_size` is stored inside `read_command` below. It has another field following it - `row_limit_high_bits`. I'm not sure but if I had to guess, extending this struct is not safe and will break mixed clusters.

Reply:
(unless our idl mechanism is clever and checks the size of each field, extending it or truncating if necessary?)

Reply:
So if I understand correctly, you want to change the meaning of the fields here. Is that correct, @denesb?

New page_size field would be used by normal queries and reversing readers that implement reversing natively and soft/hard limits will be used by unpaged queries and reversing reader

Reply:
I wish you have made this comment on the issue when Tomek proposed a new unlimited limit. Would have saved me some work.

Reply:
So basically we would have 3 fields: page_size, soft_limit and hard_limit. For normal queries all 3 are the same. Unpaged queries ignore page_size. Reverse queries on old replicas ignore page_size. Reverse queries on new replicas use page_size for efficient reverse readers and soft/hard limits for reversing reader. Is that the idea @denesb ?

Reply:
The only problem I see with this is when on old coordinator sends a limit without page_size and a new replica gets it. With new coordinator and page_size being filled in, the new replica would use page_size for all the readers but the reversing reader but when it gets limit from the old coordinator then all readers have to use the old soft/hard limits. This will create a mess unless we for example drop support for soft limit.

Reply:
> (unless our idl mechanism is clever and checks the size of each field, extending it or truncating if necessary?)

IDL stores the size of all non-native types. So it will work for structs, but changing the size of an int will not work (hence the ugly limit composed of 2 32bit ints).

Reply:
> I wish you have made this comment on the issue when Tomek proposed a new unlimited limit. Would have saved me some work.

Sorry about that. I didn't think of this until I saw the code.

Reply:
> So basically we would have 3 fields: page_size, soft_limit and hard_limit. For normal queries all 3 are the same. Unpaged queries ignore page_size. Reverse queries on old replicas ignore page_size. Reverse queries on new replicas use page_size for efficient reverse readers and soft/hard limits for reversing reader. Is that the idea @denesb ?

Yes, but we don't need to special-case their values based on the query type anymore, because we have as many fields as we have limits to transfer now. page size always means page size and soft/hard limit always mean soft/hard limit for non-native reverse/unpaged queries. Some queries might not need the soft/hard limit, but it is simpler to just always fill all fields, they take the space on the wire anyway.

Reply:
We only need to keep the old special-casing for mixed-clusters and handle the replica-side accordingly for old coordinators as well.

Reply:
> IDL stores the size of all non-native types. So it will work for structs, but changing the size of an int will not work (hence the ugly limit composed of 2 32bit ints).

So when new replica receives a `read_command` from an old coordinator, it will deserialize the two fields from the `max_result_size` struct, and the third/fourth field will be set to a default value?
And when an old replica receives a `read_command` from a new coordinator, it will truncate the received `max_result_size` struct?

Reply:
> We only need to keep the old special-casing for mixed-clusters and handle the replica-side accordingly for old coordinators as well.

So what should we set page_size to when it's missing in a read_command from an old coordinator?

Reply:
I know appending fields on an object level works because IDL stores the size of the object at the start of the serialized fields. When reading, it compares this size to the size it knows and ignores any excess data, and fills fields for whom data is missing with the specified defaults.
Actually I'm not 100% sure anymore that fields added into a nested struct are handled well. It is possible to handle them correctly with just this information, but I don't know if this is what happens. For example if IDL compares the size eagerly to the known size before reading any fields, it might draw the wrong conclusions. I guess we have to test this.

Reply:
> The only problem I see with this is when on old coordinator sends a limit without page_size and a new replica gets it. With new coordinator and page_size being filled in, the new replica would use page_size for all the readers but the reversing reader but when it gets limit from the old coordinator then all readers have to use the old soft/hard limits. This will create a mess unless we for example drop support for soft limit.

We have a cluster feature so new replica can interpret the data accordingly. Without it you are right that we'd have a problem.

Reply:
But what exactly do we do in a mixed cluster when page_size is missing?

Reply:
> > We only need to keep the old special-casing for mixed-clusters and handle the replica-side accordingly for old coordinators as well.
> 
> So what should we set page_size to when it's missing in a read_command from an old coordinator?

This same constant the coordinator is using: `query::result_memory_limiter::maximum_result_size`.
We've been using this constant for a very long time, so there is no danger of some node sending some other page-size limit. We have the ability to use custom page sizes but don't exercise it ever, except when we (and by we I mean me) decided to piggy-back the unlimited limits on top of it, which in retrospect was a terrible idea.

Reply:
I think that won't work. @tgrabiec told me that all replicas should use the same page size and picking a constant would be wrong in a mixed cluster, no?

Reply:
Actually, in a mixed cluster, we have to use hard_limit as page size even for native reverse queries because that is what old replicas will use.